cms/djangoapps/contentstore/views/helpers.py

```diff
@@ -271,6 +281,77 @@ def create_xblock(parent_locator, user, category, display_name, boilerplate=None
+    """
+    Modulestore's IdGenerator doesn't work for importing single blocks as OLX,
+    so we implement our own
+    """
+    def __init__(self, context_key):
+        super().__init__()
+        self.context_key = context_key
+
+    def create_aside(self, definition_id, usage_id, aside_type):
+        """ Generate a new aside key """
+        raise NotImplementedError()
+
+    def create_usage(self, def_id) -> UsageKey:
+        """ Generate a new UsageKey for an XBlock """
+        # Note: Split modulestore will detect this temporary ID and create a new block ID when the XBlock is saved.
+        return self.context_key.make_usage_key(def_id.block_type, LocalId())
+
+    def create_definition(self, block_type, slug=None) -> DefinitionLocator:
+        """ Generate a new definition_id for an XBlock """
+        # Note: Split modulestore will detect this temporary ID and create a new definition ID when the XBlock is saved.
+        return DefinitionLocator(block_type, LocalId(block_type))
+
+
+def import_staged_content_from_user_clipboard(parent_key: UsageKey, user):
+    """
+    Import a block (and any children it has) from "staged" OLX.
+    Does not deal with permissions or REST stuff - do that before calling this.
+
+    Returns the newly created block on success or None if the clipboard is
+    empty.
+    """
+    if not content_staging_api:
+        raise RuntimeError("The required content_staging app is not installed")
+    user_clipboard = content_staging_api.get_user_clipboard_status(user.id)
+    if (
+        not user_clipboard or
+        user_clipboard.content.status != content_staging_api.StagedContentStatus.READY
+    ):
+        # Clipboard is empty or expired/error/loading
+        return None
+    block_type = user_clipboard.content.block_type
+    olx_str = content_staging_api.get_staged_content_olx(user_clipboard.content.id)
+    node = etree.fromstring(olx_str)
+    store = modulestore()
+    with store.bulk_operations(parent_key.course_key):
+        parent_xblock = store.get_item(parent_key)
+        runtime = parent_xblock.runtime
+        # Generate the new ID:
+        id_generator = ImportIdGenerator(parent_key.context_key)
+        def_id = id_generator.create_definition(block_type, user_clipboard.source_usage_key.block_id)
+        usage_id = id_generator.create_usage(def_id)
+        keys = ScopeIds(None, block_type, def_id, usage_id)
+        # parse_xml is a really messy API. We pass both 'keys' and 'id_generator' and, depending on the XBlock, either
+        # one may be used to determine the new XBlock's usage key, and the other will be ignored. e.g. video ignores
+        # 'keys' and uses 'id_generator', but the default XBlock parse_xml ignores 'id_generator' and uses 'keys'.
+        # For children of this block, obviously only id_generator is used.
+        xblock_class = runtime.load_block_type(block_type)
+        temp_xblock = xblock_class.parse_xml(node, runtime, keys, id_generator)
```

Comment:
We should probably add some error handling here: I have an older version of the demo course with the first video pointing to [this video](https://www.youtube.com/watch?v=b7xgknqkQk8) (which is private on YouTube). Copying works fine, but when I try to paste the component, I get an enigmatic error: in Studio:
```
**Studio's having trouble saving your work**

This may be happening because of an error with our server or your internet connection. Try refreshing the page or making sure you are online.
```

The real error is:
```python
2023-04-24 16:23:15,622 ERROR 4569 [django.request] [user None] [ip None] log.py:224 - Internal Server Error: /xblock/
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/http.py", line 40, in inner
    return func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py", line 55, in parse_json_into_request
    return view_function(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/block.py", line 258, in xblock_handler
    return _create_block(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py", line 55, in parse_json_into_request
    return view_function(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/block.py", line 723, in _create_block
    created_xblock = import_staged_content_from_user_clipboard(parent_key=usage_key, user=request.user)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/helpers.py", line 342, in import_staged_content_from_user_clipboard
    temp_xblock = xblock_class.parse_xml(node, runtime, keys, id_generator)
  File "/edx/app/edxapp/edx-platform/xmodule/video_block/video_block.py", line 691, in parse_xml
    video.edx_video_id = video.import_video_info_into_val(
  File "/edx/app/edxapp/edx-platform/xmodule/video_block/video_block.py", line 1040, in import_video_info_into_val
    edx_video_id = edxval_api.import_from_xml(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edxval/api.py", line 1142, in import_from_xml
    edx_video_id = create_video(data)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edxval/api.py", line 114, in create_video
    raise ValCannotCreateError(serializer.errors)
edxval.exceptions.ValCannotCreateError: {'client_video_id': [ErrorDetail(string='This field may not be null.', code='null')], 'duration': [ErrorDetail(string='This field may not be null.', code='null')]}
```

Reply:
What sort of error handling do you have in mind? I think generally it's not a good idea to pass exceptions on to users, because they can potentially leak internal details or user PII. Or do you just want a more specific error like "There was a problem pasting that video component."

Reply:
@bradenmacdonald, yep, I thought about anything more specific than "This may be happening because of an error with our server or your internet connection. Try refreshing the page or making sure you are online.".  Something like "There was a problem pasting your component." should be OK.

Reply:
OK, I found out how to override part of the error message that's shown, so here's how it looks now:

![Screenshot 2023-04-25 at 10 42 52 AM](https://user-images.githubusercontent.com/945577/234358973-e8930c3f-4c3c-4d5b-bfda-3c53793e8567.png)

And here's the improved log:

```
2023-04-25 17:42:42,433 ERROR 177 [cms.djangoapps.contentstore.views.block] [user 4] [ip 172.18.0.1]
block.py:712 - Could not paste component into location block-v1:BradenX+CP200+23+type@vertical+block@8b69a40fac2945f9a9ddc6da197923ff
Traceback (most recent call last):
  File "/openedx/edx-platform/cms/djangoapps/contentstore/views/block.py", line 710, in _create_block
    created_xblock = import_staged_content_from_user_clipboard(parent_key=usage_key, user=request.user)
  ...
  File "/openedx/venv/lib/python3.8/site-packages/xblock/runtime.py", line 1125, in service
    raise NoSuchServiceError(f"Service {service_name!r} is not available.")
xblock.exceptions.NoSuchServiceError: Service 'field-data' is not available.
```

See db4eb973047d75b415c11a471b9072d60d7f16ae